simphony_mayavi/sources/slim_cuds_source.py

```diff
@@ -0,0 +1,623 @@
+        Instance(VTKMesh),
+        Instance(VTKParticles),
+        Instance(VTKLattice))
+
+    # This filter allows us to change the attributes of the data
+    # object and will ensure that the pipeline is properly taken care
+    # of.  Directly setting the array in the VTK object will not do
+    # this.
+    _assign_attribute = Instance(tvtk.AssignAttribute, args=(),
+                                 allow_none=False)
+
+    # The ID of the observer for the data.
+    _observer_id = Int(-1)
+
+    def __init__(self, cuds=None, point_scalars=None, point_vectors=None,
+                 cell_scalars=None, cell_vectors=None, **traits):
+        """ Constructor
+
+        Parameters
+        ----------
+        cuds : ABCParticles, ABCLattice, ABCMesh or H5Mesh
+            The CUDS dataset to be wrapped as VTK data source
+
+        point_scalars : str
+            CUBA name of the data to be selected as point scalars.
+            Default is the first available point scalars.
+
+        point_vectors : str
+            CUBA name of the data to be selected as point vectors.
+            Default is the first available point vectors.
+
+        cell_scalars : str
+            CUBA name of the data to be selected as cell scalars.
+            Default is the first available cell scalars.
+
+        cell_vectors : str
+            CUBA name of the data to be selected as cell vectors.
+            Default is the first available cell vectors.
+
+        Notes
+        -----
+        To turn off visualisation for a point/cell scalar/vector data,
+        assign the attribute to an empty string (i.e. point_scalars="")
+
+        Other optional keyword parameters are parsed to VTKDataSource.
+
+        Examples
+        --------
+        >>> cuds = Particles("test")
+
+        >>> # Say each particle has scalars "TEMPERATURE" and "MASS"
+        >>> # and vector data: "VELOCITY"
+        >>> cuds.add_particles([...])
+
+        >>> # Initialise the source and specify scalar data to visualise
+        >>> # but turn off the visualisation for point vectors
+        >>> source = SlimCUDSSource(
+                        cuds=cuds,
+                        point_scalars="MASS",
+                        point_vectors="")
+
+        >>> # Show it in Mayavi!
+        >>> from mayavi import mlab
+        >>> mlab.pipeline.glyph(source)
+        """
+        super(SlimCUDSSource, self).__init__(**traits)
+
+        # I don't want to use locals(). Too risky.
+        arguments = {
+            "point_scalars": point_scalars,
+            "point_vectors": point_vectors,
+            "cell_scalars": cell_scalars,
+            "cell_vectors": cell_vectors
+        }
+
+        defaults = {}
+        for data_type_attr in ["{}_{}".format(data_type, attr)
+                               for data_type, attr in data_type_attrs()]:
+            defaults[data_type_attr] = arguments[data_type_attr]
+
+        self._constructor_default_names = defaults
```

Comment:
ok... this is probably true. I did change after change to that part until it became degenerate without realising it.